core/src/service/sync_service.rs

```diff
@@ -24,122 +28,228 @@ pub struct WorkCalculated {
+                match op {
+                    SyncOperation::PullMetadataStart => {
+                        sync_progress.current_work_unit = ClientWorkUnit::PullMetadata;
+                    }
+                    SyncOperation::PushMetadataStart(_) => {
+                        sync_progress.current_work_unit = ClientWorkUnit::PushMetadata;
+                    }
+                    SyncOperation::PullDocumentStart(file) => {
+                        sync_progress.current_work_unit = ClientWorkUnit::PullDocument(file.name);
+                    }
+                    SyncOperation::PushDocumentStart(file) => {
+                        sync_progress.current_work_unit = ClientWorkUnit::PushDocument(file.name);
+                    }
+                    SyncOperation::PullMetadataEnd(_)
+                    | SyncOperation::PushMetadataEnd
+                    | SyncOperation::PullDocumentEnd
+                    | SyncOperation::PushDocumentEnd => {
+                        sync_progress.progress += 1;
+                    }
                 }
-                sync_progress += 1;
+                update_sync_progress(sync_progress.clone());
             }
         };
 
-        self.prune()?;
-        self.pull(&mut update_sync_progress)?;
-        self.push_metadata(&mut update_sync_progress)?;
-        self.prune()?;
-        self.push_documents(&mut update_sync_progress)?;
-        let update_as_of = self.pull(&mut update_sync_progress)?;
-        self.tx.last_synced.insert(OneKey {}, update_as_of);
-        self.populate_public_key_cache()?;
+        let mut sync_context = SyncContext {
+            dry_run: false,
+            account: self.get_account()?.clone(),
+            root: self.tx.root.get(&OneKey {}).cloned(),
+            last_synced: self.tx.last_synced.get(&OneKey {}).map(|s| *s as u64),
+            base: (&mut self.tx.base_metadata).to_staged(Vec::new()),
+            local: (&mut self.tx.local_metadata).to_staged(Vec::new()),
+            username_by_public_key: &mut self.tx.username_by_public_key,
+            public_key_by_username: &mut self.tx.public_key_by_username,
+            client: self.client,
+            config: self.config,
+        };
 
-        Ok(())
+        let update_as_of = sync_context.sync(&mut update_sync_progress)?;
+
+        if let Some(root) = sync_context.root {
+            self.tx.root.insert(OneKey {}, root);
+        }
+        if let Some(last_synced) = sync_context.last_synced {
+            self.tx.last_synced.insert(OneKey {}, last_synced as i64);
+        }
+        sync_context.base.to_lazy().promote();
+        sync_context.local.to_lazy().promote();
+
+        Ok(WorkCalculated { work_units, most_recent_update_from_server: update_as_of as u64 })
+    }
+}
+
+struct SyncContext<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+where
+    Base: TreeLike<F = SignedFile>,
+    Local: TreeLike<F = SignedFile>,
```

Comment:
note: sync does not actually modify the trees, but stages changes so that they can be promoted